agent/proxycfg/state.go

```diff
@@ -926,10 +933,53 @@ func (s *state) handleUpdateUpstreams(u cache.UpdateEvent, snap *ConfigSnapshotU
 			return fmt.Errorf("invalid correlation id %q", u.CorrelationID)
 		}
 
-		if _, ok := snap.WatchedUpstreamEndpoints[svc]; !ok {
-			snap.WatchedUpstreamEndpoints[svc] = make(map[string]structs.CheckServiceNodes)
+		if _, ok := upstreamsSnapshot.WatchedUpstreamEndpoints[svc]; !ok {
+			upstreamsSnapshot.WatchedUpstreamEndpoints[svc] = make(map[string]structs.CheckServiceNodes)
+		}
+		upstreamsSnapshot.WatchedUpstreamEndpoints[svc][targetID] = resp.Nodes
+
+		var passthroughAddrs map[string]ServicePassthroughAddrs
+
+		for _, node := range resp.Nodes {
+			if snap.Proxy.Mode == structs.ProxyModeTransparent && node.Service.Proxy.TransparentProxy.DialedDirectly {
+				if passthroughAddrs == nil {
+					passthroughAddrs = make(map[string]ServicePassthroughAddrs)
+				}
+
+				svc := node.Service.CompoundServiceName()
+
+				// Overwrite the name if it's a connect proxy (as opposed to Connect native).
+				// We don't reference the proxy name directly for things like SNI, but rather the name
+				// of the destination. The enterprise meta of a proxy will always be the same as that of
+				// the destination service, so that remains intact.
+				if node.Service.Kind == structs.ServiceKindConnectProxy {
+					dst := node.Service.Proxy.DestinationServiceName
+					if dst == "" {
```

Comment:
isn't this backwards? I thought IDs were more specific than Names

Reply:
I think we want to prefer the name for the SNI, and only fall back if it doesn't exist.

Follows this: https://github.com/hashicorp/consul/blob/master/agent/structs/structs.go#L952